### PR TITLE
Feature/more performant number grammar

### PR DESCRIPTION
**Pull request by sfgeorge**
Make the number DTMF grammar more performant for LumenVox.  Note that the more re-use we employ with rulerefs, the fewer cycles that their FST has to create.

Opening this for peer review along with the open source community.

### CODE REVIEW

**Comment by sfgeorge** [comment]
@gfaza Can you review this for me?

**Comment by benlangfeld** [comment]
Does Lumenvox not have their own pare-compiled built-in grammar for this? These grammars were intended for use with Punchblock's recogniser.

**Comment by sfgeorge** [comment]
tl;dr Here's a an attempt at a rationale for this PR.  Let me know whether or not this sounds reasonable and if so I'm happy to make the little tweaks to make hound-ci happy.  Thanks man!

> Does Lumenvox not have their own pare-compiled built-in grammar for this? These grammars were intended for use with Punchblock's recogniser.

Good question there. LumenVox does have a builtin for this, but their builtins are compiled and cached on demand, and are optimized no more than any custom grammar that you pass in (this is good).  So one's custom grammars are optimized and cached with the same priority as their own builtins, which is nice.

One benefit for using a ruby_speech number dtmf grammar is that for customers who sometimes use ASR and sometimes use DTMF-only (such as myself), there is a consistent number dtmf grammar being used.

Additionally, having more control over your grammars, such as when you use ruby_speech, is especially helpful when there are issues with the builtins provided by vendors.  Due to performance problems I encountered, I can share that LumenVox support has recommended to me to _stop using_ their builtin number grammar.  They encourage customers to use finite sequences rather than infinite whenever possible.  Their builtin number grammar itself is currently inifite and can cause trouble.  Now, this PR continues to support an infinite series for a number question, but there are 2 reasons why this is still helpful:
1. The key component to this PR is the fact that the "digit_series" ruleref is re-used.  One might think that the level of abstraction added by rulerefs would slow down a parser.  In the case of LumenVox, it is quite the opposite - the parser that generates a FST (Finite State Transducer / Finite State Automaton) only has to analyze a given ruleref _once_, and subsequent references to that same ruleref are recognized to be the same, preventing the state tree from growing even further.  This change alone makes the number grammar perform much better under LumenVox, all without changing 
2.  The API for `RubySpeech::GRXML::Builtins.number` and its use in Punchblock have superb support for passing in future options such as maxlength, much like to how `RubySpeech::GRXML::Builtins.decimal` already supports maxlength/minlength.  It would be straightforward in the future to mitigate parser issues by adding options to specify a finite limit on the integer length and decimal length of a number grammar.  The ideal length limit will likely vary widely per customer or application - and even more so why an API-powered grammar builder (achem, ruby_speech) is well-suited to solving this problem rather than a global and untouchable builtin.

To be fair, LumenVox is not the only vendor who recommends/steers toward a finite limit on a number sequence.  As seen in the [Voxeo docs](http://help.voxeo.com/go/help/xml.vxml.grxmlgram.builtin):
> **number** - Specifies a voice or DTMF input grammar that recognizes numeric input. For Prophecy ASR, the maximum length of detectable digits is 16 when using this built-in grammar.

_side note: Rebasing just to make sure that tests still pass following the nokogiri 1.7.0 release.  No recent changes to this PR._

File: lib/ruby_speech/grxml/builtins.rb [hunk 1]
```diff
@@ -116,20 +116,40 @@ def self.currency(options = nil)
   # @return [RubySpeech::GRXML::Grammar] a grammar for interpreting a numeric value.
   #
   def self.number(options = nil)
-    RubySpeech::GRXML.draw mode: :dtmf, root: 'number' do
-      rule id: 'number', scope: 'public' do
-        item repeat: '0-' do
-          ruleref uri: '#digit'
+    RubySpeech::GRXML.draw mode: :dtmf, root: "number" do
```
Comment (sfgeorge) [inline lib/ruby_speech/grxml/builtins.rb]: Good call!

File: lib/ruby_speech/grxml/grammar.rb [hunk 2]
```diff
@@ -116,9 +116,18 @@ def inline
...
-        xpath("//ns:ruleref", :ns => GRXML_NAMESPACE).each do |ref|
-          rule = rule_with_id ref[:uri].sub(/^#/, '')
-          ref.swap rule.dup.children
+        loop do
+          rule = nil
+          xpath("//ns:ruleref", ns: GRXML_NAMESPACE).each do |ref|
+            rule = rule_with_id ref[:uri].sub(/^#/, "")
+            unless rule
+              raise
+                ArgumentError,
```
Comment (sfgeorge) [inline lib/ruby_speech/grxml/grammar.rb]: > please enlighten me as to why that is

Heh, I'm really not sure why this triggers `dynamic constant assignment`.  Unfortunately, the following proper argument alignment will not suffice because it causes the line with the 2nd argument to exceed 80 characters:

```ruby
          unless rule
            raise ArgumentError,
                  "The Ruleref \"#{ref[:uri]}\" is referenced but not defined"
          end
```

I was then optimistic that hound-ci would accept non-aligned arguments based on the following rule exception in ruby style guide that it is based off of:

> Align the parameters of a method call if they span more than one line. **When aligning parameters is not appropriate due to line-length constraints, single indent for the lines after the first is also acceptable.**
https://github.com/bbatsov/ruby-style-guide#no-double-indent

But alas, hound-ci does not accept this either:

```ruby
          unless rule
            raise ArgumentError,
              "The Ruleref \"#{ref[:uri]}\" is referenced but not defined"
          end
```

In lieu of hound-ci honoring that exception to the rule, I believe I'll need to assign `"The Ruleref \"#{ref[:uri]}\" is referenced but not defined"` to a separate variable in order to appease.

File: spec/ruby_speech/grxml/builtins_spec.rb [hunk 3]
```diff
@@ -122,19 +122,23 @@
...
-    it { should match('*59').and_interpret_as('dtmf-star dtmf-5 dtmf-9') }
-    it { should match('0*0').and_interpret_as('dtmf-0 dtmf-star dtmf-0') }
-    it { should match('10*5').and_interpret_as('dtmf-1 dtmf-0 dtmf-star dtmf-5') }
-    it { should match('123*').and_interpret_as('dtmf-1 dtmf-2 dtmf-3 dtmf-star') }
-    it { should match('123*2342').and_interpret_as('dtmf-1 dtmf-2 dtmf-3 dtmf-star dtmf-2 dtmf-3 dtmf-4 dtmf-2') }
-
-    it { should not_match('#') }
+    it { should match("0").and_interpret_as("dtmf-0") }
+    it { should match("123").and_interpret_as("dtmf-1 dtmf-2 dtmf-3") }
+    it { should match("1*01").and_interpret_as("dtmf-1 dtmf-star dtmf-0 dtmf-1") }
```
Comment (sfgeorge) [inline spec/ruby_speech/grxml/builtins_spec.rb]: Great ideas for saving space, going to give these a shot, thanks man!

**Comment by sfgeorge** [comment]
Rebasing in order to get an accurate code coverage delta.

**Comment by sfgeorge** [comment]
@benlangfeld What are your thoughts on this PR?  Does my reasoning sound legit? 😁

**Comment by sfgeorge** [comment]
@benlangfeld Is it okay to merge this?

**Comment by sfgeorge** [comment]
Thank you!
